File: go/queue/circular_queue/circular_queue.go
package circular_queue

import "errors"

//循环队列实现（基于数组）

type CircularQueue struct {
	data     []interface{}
	head     int
	tail     int
	capacity int
	size     int
}

func New(capacity int) *CircularQueue {
	return &CircularQueue{
		data:     make([]interface{}, capacity),
		head:     0,
		tail:     0,
		capacity: capacity,
		size:     0,
	}
}

//入队
func (q *CircularQueue) Enqueue(val interface{}) error {
	if q.Full() {
		return errors.New("queue is full")
	}

	q.data[q.tail] = val
	q.tail = (q.tail + 1) % q.capacity
	q.size++
	return nil
}

//出队
func (q *CircularQueue) Dequeue() interface{} {
	if q.Empty() {
		return nil
	}

	res := q.data[q.head]
	q.head = (q.head + 1) % q.capacity
	q.size--
	return res
}

//队列头元素
func (q *CircularQueue) Front() interface{} {
	return q.data[q.head]
}

//队列尾部元素
func (q *CircularQueue) Rear() interface{} {
	tail := q.tail - 1
	if tail < 0 {
		tail = q.capacity - 1
	}
	return q.data[tail]
}

//队列是否为空
func (q *CircularQueue) Empty() bool {
	return q.size == 0
}

//队列是否已满
func (q *CircularQueue) Full() bool {
	return q.size == q.capacity
}

File: python/queue/circular_queue.py

# 实现一个循环队列

class CircularQueue:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.data = []
        self.size = 0
        self.head = 0
        self.tail = 0

    # 入队列
    def enque(self, val: int):
        if self.isFull():
            raise Exception('the queue is full')

        self.data.append(val)
        self.tail = (self.tail + 1) % self.capacity
        self.size += 1

    # 出队列
    def deque(self) -> int:
        if self.isEmpty():
            raise Exception('the queue is empty')

        res = self.data[self.head]
        self.head = (self.head + 1) % self.capacity
        self.size -= 1
        return res

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size >= self.capacity


if __name__ == '__main__':
    queue = CircularQueue(5)
    queue.enque(1)
    queue.enque(4)
    queue.enque(2)
    queue.enque(5)

    print(queue.deque())
    print(queue.deque())
    print(queue.deque())
    print(queue.deque())
    # print(queue.deque())
